src/core/shape_inference/include/nms_shape_inference.hpp

```diff
@@ -60,16 +66,23 @@ void shape_infer(const NonMaxSuppression* op, const std::vector<T>& input_shapes
     // NonMaxSuppression produces triplets
     // that have the following format: [batch_index, class_index, box_index]
     ov::PartialShape out_shape = {Dimension::dynamic(), 3};
-    if (boxes_ps.rank().is_static() && scores_ps.rank().is_static() && op->get_input_size() > 2) {
+    if (boxes_ps.rank().is_static() && scores_ps.rank().is_static()) {
         const auto num_boxes_boxes = boxes_ps[1];
-        if (num_boxes_boxes.is_static() && scores_ps[0].is_static() && scores_ps[1].is_static() &&
-            has_and_set_equal_bounds(op->input_value(2))) {
+        if (num_boxes_boxes.is_static() && scores_ps[0].is_static() && scores_ps[1].is_static()) {
             const auto num_boxes = num_boxes_boxes.get_length();
             const auto num_classes = scores_ps[1].get_length();
-            const auto max_output_boxes_per_class = op->max_boxes_output_from_input();
-
-            out_shape[0] =
-                Dimension(0, std::min(num_boxes, max_output_boxes_per_class) * num_classes * scores_ps[0].get_length());
+            std::vector<int64_t> max_output_boxes_per_class_as_vals;
+            bool data_available = get_data_as_int64<T>(2, op, max_output_boxes_per_class_as_vals, constant_data) ||
+                                  (op->get_input_size() > 2 &&
+                                   get_data_as_int64<T>(2, op, max_output_boxes_per_class_as_vals, constant_data));
+            if (data_available) {
```

Comment:
```suggestion
            if (op->get_input_size() > 2 && get_data_as_int64<T>(2, op, max_output_boxes_per_class_as_vals, constant_data)) {
```

Reply:
In gpu plugin, only constant data is used w/o op's input, So the condition is merged into `get_data_as_int64` as discussed.